namesys/pubsub.go

```diff
@@ -0,0 +1,346 @@
+		cr:   cr,   // needed for pubsub bootstrap
+		pkf:  pkf,
+		ps:   ps,
+		subs: make(map[string]*floodsub.Subscription),
+	}
+}
+
+// Publish publishes an IPNS record through pubsub with default TTL
+func (p *PubsubPublisher) Publish(ctx context.Context, k ci.PrivKey, value path.Path) error {
+	return p.PublishWithEOL(ctx, k, value, time.Now().Add(DefaultRecordTTL))
+}
+
+// PublishWithEOL publishes an IPNS record through pubsub
+func (p *PubsubPublisher) PublishWithEOL(ctx context.Context, k ci.PrivKey, value path.Path, eol time.Time) error {
+	id, err := peer.IDFromPrivateKey(k)
+	if err != nil {
+		return err
+	}
+
+	_, ipnskey := IpnsKeysForID(id)
+
+	seqno, err := p.getPreviousSeqNo(ctx, ipnskey)
+	if err != nil {
+		return err
+	}
+
+	seqno++
+
+	return p.publishRecord(ctx, k, value, seqno, eol, ipnskey, id)
+}
+
+func (p *PubsubPublisher) getPreviousSeqNo(ctx context.Context, ipnskey string) (uint64, error) {
+	// the datastore is shared with the routing publisher to properly increment and persist
+	// ipns record sequence numbers.
+	prevrec, err := p.ds.Get(dshelp.NewKeyFromBinary([]byte(ipnskey)))
+	if err != nil {
+		if err == ds.ErrNotFound {
+			// None found, lets start at zero!
+			return 0, nil
+		}
+		return 0, err
+	}
+
+	prbytes, ok := prevrec.([]byte)
+	if !ok {
+		return 0, fmt.Errorf("unexpected type returned from datastore: %#v", prevrec)
+	}
+
+	var dsrec dhtpb.Record
+	err = proto.Unmarshal(prbytes, &dsrec)
+	if err != nil {
+		return 0, err
+	}
+
+	var entry pb.IpnsEntry
+	err = proto.Unmarshal(dsrec.GetValue(), &entry)
+	if err != nil {
+		return 0, err
+	}
+
+	return entry.GetSequence(), nil
+}
+
+func (p *PubsubPublisher) publishRecord(ctx context.Context, k ci.PrivKey, value path.Path, seqno uint64, eol time.Time, ipnskey string, ID peer.ID) error {
+	entry, err := CreateRoutingEntryData(k, value, seqno, eol)
+	if err != nil {
+		return err
+	}
+
+	data, err := proto.Marshal(entry)
+	if err != nil {
+		return err
+	}
+
+	// the datastore is shared with the routing publisher to properly increment and persist
+	// ipns record sequence numbers; so we need to Record our new entry in the datastore
+	dsrec, err := record.MakePutRecord(k, ipnskey, data, true)
+	if err != nil {
+		return err
+	}
+
+	dsdata, err := proto.Marshal(dsrec)
+	if err != nil {
+		return err
+	}
+
+	p.ds.Put(dshelp.NewKeyFromBinary([]byte(ipnskey)), dsdata)
+
+	// now we publish, but we also need to bootstrap pubsub for our messages to propagate
+	topic := "/ipns/" + ID.Pretty()
+
+	p.mx.Lock()
+	_, ok := p.subs[topic]
+
+	if !ok {
+		p.subs[topic] = struct{}{}
+		p.mx.Unlock()
+
+		bootstrapPubsub(p.ctx, p.cr, p.host, topic)
+	} else {
+		p.mx.Unlock()
+	}
+
+	log.Debugf("PubsubPublish: publish IPNS record for %s", topic)
+	return p.ps.Publish(topic, data)
+}
+
+// Resolve resolves a name through pubsub and default depth limit
+func (r *PubsubResolver) Resolve(ctx context.Context, name string) (value path.Path, err error) {
```

Comment:
removed the usage of named values in https://github.com/ipfs/go-ipfs/pull/4047/commits/62e427c1b5284c5f75f5ee774338df182b0493c2